lib/ecto/schema.ex

```diff
@@ -1517,7 +1518,8 @@ defmodule Ecto.Schema do
   defmacro embeds_one(name, schema, opts, do: block) do
     quote do
       {schema, opts} = Ecto.Schema.__embeds_module__(__ENV__, unquote(schema), unquote(opts), unquote(Macro.escape(block)))
-      Ecto.Schema.__embeds_one__(__MODULE__, unquote(name), schema, opts)
+      opts = Keyword.put(opts, :type, unquote(schema))
```

Comment:
We should make this a feature of the parameterized types too (i.e. they all receive the current schema as an option).

Reply:
💯  I'll inject that.